codecombat/codecombat
# Initial Functional Character Customization Screen

## Features

Initial functional page of player character customization, with some minor cinematic changes so that the customization propagates into the cinematic.

- User can navigate to http://localhost:3000/ozaria/character-customization and see a character customization page.
- User can select a body, name and tints. These will be reflected in cinematics. (Not levels yet).
- If user has previously saved options, the form should be pre-populated, otherwise randomly selected.

## How to manually test and verify this change

1. Start a local proxied environment by running `npm run proxy` and `npm run dev` simultaneously.
1. Log in with your admin account
1. Navigate to http://localhost:3000/ozaria/character-customization
1. Character customize and press Next. (Your choices are saved!)
1. Navigate to http://localhost:3000/cinematic/1fhl1c1 and the right hand character should reflect the customized character.

## Risks

I believe the risks of shipping this are minimal as the vast majority of code paths are behind admin flags. There is no high risk code changes here as the color tinting system changes are already in production. This PR provides an interface to set the character customization data.

## Known Issues

 - When the user presses Next there is a validation error. However the information still saves.
    - I believe this is due to my own user state as I can repro the error everywhere on my account without any of the character customization changes.
 - Art isn't yet set up to support full customization. But you can test it with Hero A.
 - Various error handling locations and UI need improvement.
 - This page will need to be moved into a modal.
 - Component level tests were neglected due to time. The api's are tested.
 - All api calls could not be moved into Vuex due to time as well.


_Notice in images below the hair and skin highlight persists into cinematic_
<img width="1563" alt="Screen Shot 2019-07-12 at 10 21 35 AM" src="https://user-images.githubusercontent.com/15080861/61146607-49fb5f00-a48f-11e9-8e38-0e28ff0d8655.png">

<img width="1563" alt="Screen Shot 2019-07-12 at 10 21 10 AM" src="https://user-images.githubusercontent.com/15080861/61146608-49fb5f00-a48f-11e9-8f66-f3a94c6db1b5.png">
